src/replication.c

```diff
@@ -2526,6 +3152,143 @@ int replicaTryPartialResynchronization(connection *conn, int read_reply) {
     return PSYNC_NOT_SUPPORTED;
 }
 
+/* Replication: Replica side.
+ * This connection handler fires after rdb-connection was initialized. We use it
+ * to adjust the replica main for loading incremental changes into the local buffer. */
+void setupMainConnForPsync(connection *conn) {
+    int psync_result = -1;
+    char llstr[LONG_STR_SIZE];
+    char *err = NULL;
+    if (server.repl_state == REPL_STATE_SEND_HANDSHAKE) {
+        /* We already have an initialized connection at primary side, we only need to associate it with RDB connection */
+        ll2string(llstr, sizeof(llstr), server.rdb_client_id);
+        err = sendCommand(conn, "REPLCONF", "set-rdb-client-id", llstr, NULL);
+        if (err) goto error;
+        server.repl_state = REPL_STATE_RECEIVE_CAPA_REPLY;
+        return;
+    }
+
+    if (server.repl_state == REPL_STATE_RECEIVE_CAPA_REPLY) {
+        err = receiveSynchronousResponse(conn);
+        if (err == NULL) goto error;
+        if (err[0] == '-') {
+            serverLog(LL_NOTICE, "Primary does not understand REPLCONF identify: %s", err);
+            goto error;
+        }
+        sdsfree(err);
+        err = NULL;
+        server.repl_state = REPL_STATE_SEND_PSYNC;
+    }
+
+    if (server.repl_state == REPL_STATE_SEND_PSYNC) {
+        if (server.debug_sleep_after_fork_us) usleep(server.debug_sleep_after_fork_us);
+        if (replicaTryPartialResynchronization(conn, 0) == PSYNC_WRITE_ERROR) {
+            serverLog(LL_WARNING, "Aborting dual connection sync. Write error.");
+            cancelReplicationHandshake(1);
+        }
+        server.repl_state = REPL_STATE_RECEIVE_PSYNC_REPLY;
+        return;
```

Comment:
not a blocker to this PR but the mixed use of `goto` and `return` makes it hard to track memory allocations and easy to leak. I am proposing a new idiom that declares all memory that need to be freed at the top of the function and a fixed `cleanup` label at the end of the function, which covers both success and failure cases and frees all memory allocations. This idiom has worked really well for me in many large code bases without the help of RAII. @valkey-io/core-team to chime in.

```c
int func() {
   void *a = NULL,*b = NULL,*c = NULL;
  
   if (...) {
       ...
       a = zmalloc();
       ...
   }
   
   if (...) {
       ...
       b = zmalloc();
       ...
   }   

    if (...) {
        ...
        goto cleanup;
        ....
    }

cleanup:
    zfree(a);
    zfree(b);
    zfree(c);

    return C_OK or C_ERR;
}
```